### DISCUSSION TITLE
StorageFactory error

### DISCUSSION THREAD

**Discussion by GenevaRoc** [Cyfrin/foundry-full-course-cu#801] [Q&A] [answered]
my code for the storage factory compiled fine but when I test it, sfStore  throws an error "invalid big number string"
here is my code
.......
  function sfStore(uint256 _simpleStorageIndex, uint256 _newSimpleStorageNumber) public{
        SimpleStorage mySimpleStorage = listOfSimpleStorageContracts[_simpleStorageIndex];
        mySimpleStorage.store(_newSimpleStorageNumber);
    }

    function sfGet(uint256 _simpleStorageIndex) public view returns(uint256){
        SimpleStorage mySimpleStorage = listOfSimpleStorageContracts[_simpleStorageIndex];
        return mySimpleStorage.retrieve();
    }

**Comment by theirrationalone** [comment]
Try to mitigate with this [comment](https://github.com/Cyfrin/foundry-full-course-f23/discussions/706#discussioncomment-6815363)

**Reply by GenevaRoc** [reply]
What do you mean?

**Reply by theirrationalone** [reply]
Click on link embedded into text `comment` then follow afterwards...

**Reply by GenevaRoc** [reply]
it still throws the error 'invalid bigNumber string'
[remix error.pdf](https://github.com/Cyfrin/foundry-full-course-f23/files/12539995/remix.error.pdf)

**Reply by theirrationalone** [reply]
Would you like to share your full code, `SimpleStorage` as well as `StorageFactory`?? So I will also be able to test it. If we're missing something 😁

**Reply by GenevaRoc** [reply]
SimpleStorage code
`// SPDX-License-Identifier: MIT
pragma solidity 0.8.18;

contract SimpleStorage {
    //favouriteNumber gets initialised to 0 if no value is given
    uint256 public myfavouriteNumber;

    struct person{
        uint256 favouriteNumber;
        string name;
    }
    person[] public listOfPeople;

    mapping(string => uint256) public nameTofavouriteNumber;

    function store(uint256 _favouriteNumber) public{
        myfavouriteNumber = _favouriteNumber;
        //view function disallow updating state
        //while pure function disallow updating or reading from state
    }

    function retrieve() public view returns(uint256){
        return myfavouriteNumber;
    }

    function addPerson(string memory _name, uint256 _favouriteNumber) public{
        person memory newPerson = person(_favouriteNumber, _name);
        listOfPeople.push(newPerson);
        nameTofavouriteNumber[_name] = _favouriteNumber;
    }
}
`


StorageFactory code
`// SPDX-License-Identifier: MIT
pragma solidity 0.8.18;


import "./SimpleStorage.sol";

contract StorageFactory {
    SimpleStorage[] public listOfSimpleStorageContracts;

    function createSimpleStorageContract() public {
        SimpleStorage newSimpleStorageContract = new SimpleStorage();
        listOfSimpleStorageContracts.push(newSimpleStorageContract);

    }

    function sfStore(uint256 _simpleStorageIndex, uint256 _newSimpleStorageNumber) public{
       SimpleStorage mySimpleStorage = listOfSimpleStorageContracts[_simpleStorageIndex];
       mySimpleStorage.store(_newSimpleStorageNumber);
    }

    function sfGet(uint256 _simpleStorageIndex) public view returns(uint256){
        return listOfSimpleStorageContracts[_simpleStorageIndex].retrieve();
    }
}`

**Reply by theirrationalone** [reply]
Oh, yes, I've identified the issue.

You're actually providing a floating-point or real number as an argument to the `sfStore` function. However, `sfStore` expects an integral number, such as 0, 1, 2, and so on.

Secondly, `sfStore` expects an array index. In programming, the first index of any array is 0, and it increments from there in a positive direction.

Thirdly, it's worth noting that Solidity has limited support for floating-point or real numbers. You'll come to appreciate its significance in later lessons. 😉;`)

**Reply by GenevaRoc** [reply]
thank you
so where do i make corrections to my code?

**Reply by theirrationalone** [reply]
No where, Just pass integers to sfStore function before clicking on its orange button. In your case, I think you need to pass 0. Since you might have only one contract into your contracts list/array.

**Reply by theirrationalone** [reply]
Steps to interact with it. 

1. CreateSimpleStorage. It will push created simplestorage instance into the list... 

2. sfStore: put 0, 123 into its orange button input box and click on it. 

3. sfGet: put 0 and click on its button.

**Reply by theirrationalone** [reply]
![IMG_20230906_221907.jpg](https://github.com/Cyfrin/foundry-full-course-f23/assets/96755510/e0b68b89-c1ff-46a0-ad47-64c51e8df171)

![IMG_20230906_222200.jpg](https://github.com/Cyfrin/foundry-full-course-f23/assets/96755510/322dcde4-5ed5-49c4-8037-e4edac4c364a)

**Reply by GenevaRoc** [reply]
thank you. it works now

**Reply by theirrationalone** [reply]
Woww, You're Welcome @GenevaRoc ! ❤

**Comment by GenevaRoc** [comment]
`// SPDX-License-Identifier: MIT
pragma solidity 0.8.18;

contract SimpleStorage {
    //favouriteNumber gets initialised to 0 if no value is given
    uint256 public myfavouriteNumber;

    struct person{
        uint256 favouriteNumber;
        string name;
    }
    person[] public listOfPeople;

    mapping(string => uint256) public nameTofavouriteNumber;

    function store(uint256 _favouriteNumber) public{
        myfavouriteNumber = _favouriteNumber;
        //view function disallow updating state
        //while pure function disallow updating or reading from state
    }

    function retrieve() public view returns(uint256){
        return myfavouriteNumber;
    }

    function addPerson(string memory _name, uint256 _favouriteNumber) public{
        person memory newPerson = person(_favouriteNumber, _name);
        listOfPeople.push(newPerson);
        nameTofavouriteNumber[_name] = _favouriteNumber;
    }
}
`

st

**Reply by theirrationalone** [reply]
Sorry for this 😅. Please consider this [comment](https://github.com/Cyfrin/foundry-full-course-f23/discussions/801#discussioncomment-6927962) instead.

**Reply by GenevaRoc** [reply]
hey
i'm trying to create a new project on foundry but it says forge isn't installed. Any idea on how I can get forge installed?
this is what it says
`Command 'forge' not found, but can be installed with:
sudo apt install snap`

**Reply by theirrationalone** [reply]
Try these cmds:

```bash
foundryup
source Users/your_username/.bashrc
```

**Reply by GenevaRoc** [reply]
foundryup throws an error, am I supposed to add a command before it?

**Reply by theirrationalone** [reply]
What that error is?? 

Or You may need to watch patrick's foundry installation method again.

**Reply by GenevaRoc** [reply]
here is the error I'm getting
`Error: 
failed to commit (code=Some(128), stdout="", stderr="Author identity unknown\n\n*** Please tell me who you are.\n\nRun\n\n  git config --global user.email \"you@example.com\"\n  git config --global user.name \"Your Name\"\n\nto set your account's default identity.\nOmit --global to set the identity only in this repository.\n\nfatal: empty ident name (for <genevaroc@DESKTOP-FJAPMMJ.>) not allowed")`

**Reply by theirrationalone** [reply, marked as answer]
Please tell github who are u... 

Execute command below: (in my case) 
Your github registered email:
```bash
git config --global user.email theirrationalone@example.com
```
Your github username:
```bash
git config --global user.name theirrationalone
```

After that try again please. 😉

**Reply by GenevaRoc** [reply]
it worked, lol
thank you

**Reply by GenevaRoc** [reply]
hey, I'm testing on foundry but anytime i call the listOfpeople function, it throws me this error
`(code: 3, message: execution reverted, data: Some(String("0x")))`

this is the command i use
`cast call 0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0 "listOfPeople()"`

this is the actual code
`function addPerson(string memory _name, uint256 _favouriteNumber) public{
        person memory newPerson = person(_favouriteNumber, _name);
        listOfPeople.push(newPerson);
        nameTofavouriteNumber[_name] = _favouriteNumber;`

**Reply by theirrationalone** [reply]
Are you sure `listOfPeople()` isn't expecting any arg?? 

If so, Please show me your `listOfPeople()` code. 

There could have solidity limitations too but i'm not sure. 😅

**Reply by GenevaRoc** [reply]
here is the code
` struct person{
        uint256 favouriteNumber;
        string name;
    }
    person[] public listOfPeople;`

**Reply by theirrationalone** [reply]
Ohh damn, `listOfPeople` is an array no doubt an array always expects an index to point data for storing/retrieving data. 

You need to pass an integer like this below. 

```bash
cast call 0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0 "listOfPeople(uint256)" 0
```

And please make sure you have stored some data into that array by calling addPerson function first. Then you can retrieve data in sequential manner.

**Reply by GenevaRoc** [reply]
thank you

**Reply by GenevaRoc** [reply]
i'm having an error trying to initiate a fork to a node on alchemy
I think I'm doing everything right
here is the command 
`forge test --fork-url $SAPOLIA_RPC_URL`

**Reply by GenevaRoc** [reply]
here is the error I'm getting
`Could not instantiate forked environment with fork url`

**Reply by GenevaRoc** [reply]
this is the context of the error
`Error #0: Failed to get latest block number
- Error #1: (code: -32000, message: Must be authenticated!, data: None)`

**Reply by theirrationalone** [reply]
Alchemy! Your txs. Are in pending state and being stucked into a block. 

Please try to create new metamask and new alchemy rpc. 

And then try again!

**Reply by GenevaRoc** [reply]
it ran successfully!!! why was that the issue tho? can you explain it to me?

**Reply by theirrationalone** [reply]
It's been a famous issue here, imo it's an api connection lease issue. They might be working on this tho.
